metric_learn/pgdm.py

```diff
@@ -0,0 +1,436 @@
+    else:
+      self.A_ = check_array(self.A0)
+    
+    return a,b,c,d
+
+  def _fit_full(self, X, constraints):
+    """Learn full metric using PGDM.
+    Parameters
+    ----------
+    X : (n x d) data matrix
+        each row corresponds to a single instance
+    constraints : 4-tuple of arrays
+        (a,b,c,d) indices into X, with (a,b) specifying similar and (c,d)
+        dissimilar pairs
+    """
+    a,b,c,d = constraints
+    num_pos = len(a)
+    num_neg = len(c)
+    num_samples, num_dim = X.shape
+    
+    error1 = error2 = 1e10
+    eps = 0.01        # error-bound of iterative projection on C1 and C2
+    A = self.A_
+    
+    # Create weight vector from similar samples
+    pos_diff = X[a] - X[b]
+    w = np.einsum('ij,ik->jk', pos_diff, pos_diff).ravel()
+    # `w` is the sum of all outer products of the rows in `pos_diff`.
+    # The above `einsum` is equivalent to the much more inefficient:
+    # w = np.apply_along_axis(
+    #         lambda x: np.outer(x,x).ravel(),
+    #         1,
+    #         X[a] - X[b]
+    #     ).sum(axis = 0)
+    t = w.dot(A.ravel() / 100.0)
+    
+    w1 = w / np.linalg.norm(w) # make `w` a unit vector
+    t1 = t / np.linalg.norm(w) # distance from origin to `w^T*x=t` plane
+    
+    cycle = 1
+    alpha = 0.1 # initial step size along gradient
+    
+    grad1 = self._fS1(X, a, b, A)           # gradient of similarity constraint function
+    grad2 = self._fD1(X, c, d, A)           # gradient of dissimilarity constraint function
+    M = self._grad_projection(grad1, grad2) # gradient of fD1 orthogonal to fS1
+    
+    A_old = A.copy()
+
+    for cycle in xrange(self.max_iter):
+      
+      # projection of constraints C1 and C2
+      satisfy = False
+      
+      for it in xrange(self.max_proj):
+        
+        # First constraint:
+        # f(A) = \sum_{i,j \in S} d_ij' A d_ij <= t              (1)
+        # (1) can be rewritten as a linear constraint: w^T x = t,
+        # where x is the unrolled matrix of A,
+        # w is also an unrolled matrix of W where
+        # W_{kl}= \sum_{i,j \in S}d_ij^k * d_ij^l
+        x0 = A.ravel()
+        if w.dot(x0) <= t:
+          x = x0
+        else:
+          x = x0 + (t1 - w1.dot(x0)) * w1
+          A[:] = x.reshape(num_dim, num_dim)
+        
+        # Second constraint:
+        # PSD constraint A >= 0
+        # project A onto domain A>0
+        l, V = np.linalg.eigh((A + A.T) / 2)
+        A[:] = np.dot(V * np.maximum(0, l[None,:]), V.T)
+        
+        fDC2 = w.dot(A.ravel())
+        error2 = (fDC2 - t) / t
+        if error2 < eps:
+          satisfy = True
+          break
+      
+      # third constraint: gradient ascent
+      # max: g(A) >= 1
+      # here we suppose g(A) = fD(A) = \sum_{I,J \in D} sqrt(d_ij' A d_ij)
+      
+      obj_previous = self._fD(X, c, d, A_old) # g(A_old)
+      obj = self._fD(X, c, d, A)              # g(A)
+      
+      if ((obj > obj_previous) or (cycle == 0)) and (satisfy):
```

Comment:
Don't need so many parens here:

    if satisfy and (obj > obj_previous or cycle == 0):